### PR TITLE
[PLATFORM-287] Notification refactoring

### PR DESCRIPTION
**Pull request by mondoreale**
Here we go with more refactoring! The aim here is to make notifications as easy to work with as possible from anywhere, pretty much.

**The idea**
The new setup comes with 2 elements: `Notification` object and `Notifications` component. `Notifications`, when mounted, starts listening for _push_ events emitted by `Notification.push`. It subs to the event on mount and unsubs before being unmounted.

- `Notification` used to be a type. Now it's an object with functionality.
- It works nicely with modals (I ported the code @juhah wrote for it).

**Motivation behind event emitters, etc.**
Atm we trigger notifications in components, redux actions and utility functions. For regular single-result-having actions we could go with promises and go with more React-like approach for triggering notifications (via contexts, for instance). But! It does get complicated for transactions where we return a mixture of event emitter and promise. We'd have to support promise, promise-o-emitter and emitter. I decided to go with a single emitter then.

Please, go try it out. Share your thoughs. Thanks!

### CODE REVIEW

File: app/src/marketplace/containers/ProductPage/index.jsx [hunk 1]
```diff
@@ -144,13 +144,9 @@ export class ProductPage extends Component<Props, State> {
...
-    getPublishButtonDisabled = (product: Product) => {
-        if (product.state === productStates.DEPLOYING ||
-            product.state === productStates.UNDEPLOYING) {
-            return true
-        }
-        return false
-    }
+    getPublishButtonDisabled = (product: Product) => (
+        product.state === productStates.DEPLOYING || product.state === productStates.UNDEPLOYING
+    )
```
Comment (juhah) [inline app/src/marketplace/containers/ProductPage/index.jsx]: 👍

File: app/src/shared/components/Notifications/BasicNotification/basic.pcss [hunk 2]
```diff
@@ -12,7 +12,7 @@
 }
 
 .error {
-  background: url('../../../assets/error.svg');
+  background: url('../../../../marketplace/assets/error.svg');
```
Comment (juhah) [inline app/src/shared/components/Notifications/BasicNotification/basic.pcss]: You could move the image to shared as well.

File: app/src/shared/utils/NotificationIcon/index.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+// @flow
+
+export default {
```
Comment (juhah) [inline app/src/shared/utils/NotificationIcon/index.js]: I think these would be better if kept in `constants.js`.

File: app/src/shared/components/Notifications/index.jsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
+// @flow
+
+import React, { type Node } from 'react'
+import NotificationSystem from 'react-notification-system'
+
+import BasicNotification from './BasicNotification'
+import TransactionNotification from './TransactionNotification'
+import ModalContext from '$shared/contexts/Modal'
+import Notification from '$shared/utils/Notification'
+import styles from './notifications'
```
Comment (juhah) [inline app/src/shared/components/Notifications/index.jsx]: How about renaming this file to `notificationStyles`? It's not a `.pcss` file so no need to keep up the convention :)

File: app/src/shared/utils/Notification/index.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        this.txHash = params.txHash || null
+        this.icon = params.icon || null
+    }
+
+    isTx() {
+        return !!this.txHash
+    }
+
+    autoDismissAfter() {
+        return this.isTx() ? 0 : 5 // seconds, 0 = no automatic dismiss
```
[the remark this answers is not in the document]
Comment (juhah) [reply app/src/shared/utils/Notification/index.js]: I was thinking about leaving a similar comment but it's a bit out of scope from this PR. The notifications are now separated into "normal" notifications and transaction notifications mainly in order to show a correct note when the transaction is finished (eg. "Product is published"). But the transaction type should be available now when polling the completion so the notification text could be formed there.

File: app/src/shared/utils/Notification/index.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    static unsubscribe(handler: (Notification) => void) {
+        emitter.removeListener(Notification.events.PUSH, handler)
+    }
+
+    static events = events
+
+    constructor(params: Params) {
+        const now = new Date()
+        this.id = now.getTime()
```
[the remark this answers is not in the document]
Comment (mondoreale) [reply app/src/shared/utils/Notification/index.js]: I added a test that confirms your "theory" (it fails deterministically when time is frozen). I like the `++` way better altho I couldn't use the unary operator #eslint. Cheers!

File: app/src/shared/utils/Notification/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mondoreale) [reply app/src/shared/utils/Notification/index.js]: I turned `autoDismissAfter` into a property and control it using an `autoDismiss` flag (`true` unless explicitly set to `false`). I'll leave Transaction notifications out of this PR. We can refactor it later.

File: app/src/shared/components/Notifications/notificationStyles.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        info: {
+            width: 'auto',
+            borderTop: 0,
+            padding: '16px',
+            backgroundColor: 'white',
+            color: 'black',
+            WebkitBoxShadow: '0 0 10px rgba(0, 0, 0, 0.2)',
+            MozBoxShadow: '0 0 10px rgba(0, 0, 0, 0.2)',
+            boxShadow: '0 0 10px rgba(0, 0, 0, 0.2)',
```
[the remark this answers is not in the document]
Comment (mondoreale) [reply app/src/shared/components/Notifications/notificationStyles.js]: Happiness is reality minus expectations…

**Pull request merged**